### DISCUSSION TITLE
What layer to use to implement a Linear Heatmap from GeoJson

### DISCUSSION THREAD

**Discussion by a0ax** [maplibre/maplibre-gl-js#3813] [Q&A] [answered]
I'm looking to build something that takes GeoJSON input and creates a linear kind of heatmap, like the example below, using MapLibre. 
![Strava-tools-jonathan-o-keeffe](https://github.com/maplibre/maplibre-gl-js/assets/61996958/6c657f93-d10e-4acf-b6d1-2543bb167bdd)
I.e. the density of traces at a point defines the color at that point. 
**What type of layer should I be using to do this?**
Any and all help much appreciated.

**Comment by HarelM** [comment, marked as answer]
I don't think there's a built in solution for this.
The closes we got to a strava like activity heat map was using lines and style them to be a bit transparent, so that the more there are, the transparency would go away, see here:
![image](https://github.com/maplibre/maplibre-gl-js/assets/3269297/69b679f8-0fcf-4fbf-8eed-6a2c85f2ec99)
Since this is based on actual GPS recording, the more recording you have, the more noise you get and the line gets thicker too.

Not sure this is exactly what you are looking for, and there might be other people who did other stuff using CustomLayer and god know what, but it might be enough for your use case.
Can be seen in the following site:
https://israelhiking.osm.org.il/

**Reply by TimMcCauley** [reply]
And to make sure you can group the individual tracks, you could use mapmatching and then count to determine the density.

**Reply by a0ax** [reply]
Thanks both for the input. I'm hoping to be able to have separate hues for separate densities and will try to write something to achieve that, but I'll definitely keep the opacity technique as a backup.

@TimMcCauley do you think you could link me to somewhere I can read more about mapmatching? I'm not familiar with what that is.

**Reply by TimMcCauley** [reply]
@sudolev take a look at [this tutorial on medium](https://medium.com/@hernandezgalaviz/using-valhalla-map-matching-to-get-route-and-travelled-distance-from-raw-gps-points-4ea6b1c88a4c) it explains it well with examples and uses Valhalla routing which I can highly recommend as a FOSS solution to achieve it.

**Reply by a0ax** [reply]
@TimMcCauley That looks perfect, thank you so much for helping me out. Will without a doubt be helpful--I'll definitely be using valhalla.
